File: src/calculus_limits_algebraic_hard.py
"""
calculus - limits_algebraic (hard)
Generated: 2026-02-22T04:59:25.861963
"""

from problem_utils import *

def generate():
    problem_type = randint(1, 6)
    
    if problem_type == 1:
        # Limit involving factoring and cancellation with higher powers
        r1 = nonzero(-5, 5)
        r2 = nonzero(-5, 5)
        while r2 == r1:
            r2 = nonzero(-5, 5)
        
        # Numerator and denominator both have (x - r1) as a factor
        # Make it more complex with additional terms
        p = nonzero(-4, 4)
        q = nonzero(-4, 4)
        
        numerator = expand((x - r1) * (x**2 + p*x + q))
        denominator = expand((x - r1) * (x - r2))
        
        # The limit as x approaches r1
        # After canceling (x - r1), we get (x^2 + p*x + q) / (x - r2)
        ans = (r1**2 + p*r1 + q) / (r1 - r2)
        ans = simplify(ans)
        
        return problem(
            question=f"\\lim_{{x \\to {r1}}} \\frac{{{latex(numerator)}}}{{{latex(denominator)}}}",
            answer=ans,
            difficulty=(1600, 1750),
            topic="calculus/limits_algebraic",
            solution=steps(
                f"Factor the numerator: ${latex(numerator)} = {latex(factor(numerator))}$",
                f"Factor the denominator: ${latex(denominator)} = {latex(factor(denominator))}$",
                f"Cancel the common factor $(x - {r1})$",
                f"$\\lim_{{x \\to {r1}}} \\frac{{{latex(x**2 + p*x + q)}}}{{{latex(x - r2)}}}$",
                f"Substitute $x = {r1}$: $\\frac{{{r1**2 + p*r1 + q}}}{{{r1 - r2}}} = {latex(ans)}$"
            ),
        )
    
    elif problem_type == 2:
        # Limit with nested radicals requiring rationalization
        a = randint(1, 5)
        b = randint(1, 5)
        c = nonzero(-4, 4)
        
        # lim (sqrt(ax + b) - sqrt(b)) / x as x -> 0
        # Multiply by conjugate, answer is a/(2*sqrt(b))
        ans = Rational(a, 2) / sqrt(b)
        ans = simplify(ans)
        
        return problem(
            question=f"\\lim_{{x \\to 0}} \\frac{{\\sqrt{{{a}x + {b}}} - \\sqrt{{{b}}}}}{{x}}",
            answer=ans,
            difficulty=(1650, 1800),
            topic="calculus/limits_algebraic",
            solution=steps(
                f"Multiply numerator and denominator by the conjugate $\\sqrt{{{a}x + {b}}} + \\sqrt{{{b}}}$",
                f"$\\lim_{{x \\to 0}} \\frac{{({a}x + {b}) - {b}}}{{x(\\sqrt{{{a}x + {b}}} + \\sqrt{{{b}}})}}$",
                f"$\\lim_{{x \\to 0}} \\frac{{{a}x}}{{x(\\sqrt{{{a}x + {b}}} + \\sqrt{{{b}}})}}$",
                f"Cancel $x$: $\\lim_{{x \\to 0}} \\frac{{{a}}}{{\\sqrt{{{a}x + {b}}} + \\sqrt{{{b}}}}}$",
                f"Substitute $x = 0$: $\\frac{{{a}}}{{\\sqrt{{{b}}} + \\sqrt{{{b}}}}} = \\frac{{{a}}}{{{2}\\sqrt{{{b}}}}} = {latex(ans)}$"
            ),
        )
    
    elif problem_type == 3:
        # Limit involving absolute value requiring one-sided analysis
        a = nonzero(-5, 5)
        b = nonzero(-3, 3)
        
        # lim (|x - a| + b(x - a)) / (x - a) as x -> a
        # From right: (x-a + b(x-a))/(x-a) = 1 + b
        # From left: (-(x-a) + b(x-a))/(x-a) = -1 + b
        # Limit exists only if 1 + b = -1 + b, which is false
        # So we need to pick one side or make them equal
        
        # Instead, let's do a problem where limit exists
        # lim (x^2 - a^2) / |x - a| as x -> a from right
        # = lim (x+a)(x-a) / (x-a) = lim (x+a) = 2a
        
        ans = 2 * a
        
        return problem(
            question=f"\\lim_{{x \\to {a}^+}} \\frac{{x^2 - {a**2}}}{{|x - {a}|}}",
            answer=ans,
            difficulty=(1700, 1850),
            topic="calculus/limits_algebraic",
            solution=steps(
                f"Since $x \\to {a}^+$, we have $x > {a}$, so $|x - {a}| = x - {a}$",
                f"$\\lim_{{x \\to {a}^+}} \\frac{{x^2 - {a**2}}}{{x - {a}}}$",
                f"Factor: $x^2 - {a**2} = (x - {a})(x + {a})$",
                f"$\\lim_{{x \\to {a}^+}} \\frac{{(x - {a})(x + {a})}}{{x - {a}}}$",
                f"Cancel $(x - {a})$: $\\lim_{{x \\to {a}^+}} (x + {a})$",
                f"Substitute: ${a} + {a} = {ans}$"
            ),
        )
    
    elif problem_type == 4:
        # Limit with difference of cube roots
        a = randint(1, 4)
        b = randint(1, 4)
        
        # lim (cuberoot(ax + b) - cuberoot(b)) / x as x -> 0
        # Use a^3 - b^3 = (a-b)(a^2 + ab + b^2) formula
        # Answer is a / (3 * b^(2/3))
        
        ans = Rational(a, 3) / (b**Rational(2, 3))
        ans = simplify(ans)
        
        return problem(
            question=f"\\lim_{{x \\to 0}} \\frac{{\\sqrt[3]{{{a}x + {b}}} - \\sqrt[3]{{{b}}}}}{{x}}",
            answer=ans,
            difficulty=(1750, 1900),
            topic="calculus/limits_algebraic",
            solution=steps(
                f"Let $u = \\sqrt[3]{{{a}x + {b}}}$ and $v = \\sqrt[3]{{{b}}}$",
                f"Use the identity $u - v = \\frac{{u^3 - v^3}}{{u^2 + uv + v^2}}$",
                f"$u^3 - v^3 = {a}x + {b} - {b} = {a}x$",
                f"$\\lim_{{x \\to 0}} \\frac{{{a}x}}{{x(u^2 + uv + v^2)}}$",
                f"Cancel $x$: $\\lim_{{x \\to 0}} \\frac{{{a}}}{{u^2 + uv + v^2}}$",
                f"As $x \\to 0$: $u \\to \\sqrt[3]{{{b}}} = v$",
                f"$\\frac{{{a}}}{{v^2 + v \\cdot v + v^2}} = \\frac{{{a}}}{{3v^2}} = \\frac{{{a}}}{{3 \\cdot {b}^{{2/3}}}} = {latex(ans)}$"
            ),
        )
    
    elif problem_type == 5:
        # Limit requiring polynomial division/simplification
        p = nonzero(-3, 3)
        q = nonzero(-3, 3)
        
        # lim (x^3 + px^2 + qx) / (x^2 + px + q) as x -> infinity
        # Divide numerator and denominator by x^2
        # = lim (x + p + q/x) / (1 + p/x + q/x^2)
        # Answer is infinity or -infinity depending on sign
        
        # To get a finite limit, we need degree of num = degree of denom
        # lim (x^2 + px + q) / (x^2 + rx + s) as x -> infinity = 1
        
        r = nonzero(-4, 4)
        s = nonzero(-4, 4)
        
        numerator = x**2 + p*x + q
        denominator = x**2 + r*x + s
        
        # Leading coefficient ratio
        ans = 1
        
        return problem(
            question=f"\\lim_{{x \\to \\infty}} \\frac{{{latex(numerator)}}}{{{latex(denominator)}}}",
            answer=ans,
            difficulty=(1600, 1700),
            topic="calculus/limits_algebraic",
            solution=steps(
                f"Divide numerator and denominator by $x^2$ (highest power)",
                f"$\\lim_{{x \\to \\infty}} \\frac{{1 + \\frac{{{p}}}{{x}} + \\frac{{{q}}}{{x^2}}}}{{1 + \\frac{{{r}}}{{x}} + \\frac{{{s}}}{{x^2}}}}$",
                f"As $x \\to \\infty$, all terms with $x$ in denominator approach $0$",
                f"$\\frac{{1 + 0 + 0}}{{1 + 0 + 0}} = 1$"
            ),
        )
    
    else:  # problem_type == 6
        # Limit with trigonometric-like algebraic manipulation
        a = randint(2, 5)
        b = randint(1, 4)
        r = nonzero(-4, 4)
        
        # lim ((x+r)^a - r^a) / x as x -> 0
        # Using binomial or L'Hopital conceptually
        # derivative of (x+r)^a at x=0 gives a*r^(a-1)
        
        ans = a * (r**(a-1))
        
        return problem(
            question=f"\\lim_{{x \\to 0}} \\frac{{(x + {r})^{a} - {r**a}}}{{x}}",
            answer=ans,
            difficulty=(1650, 1800),
            topic="calculus/limits_algebraic",
            solution=steps(
                f"Recognize this as the definition of a derivative: $f'(0)$ where $f(x) = (x + {r})^{a}$",
                f"Alternatively, expand using binomial theorem:",
                f"$(x + {r})^{a} = {r**a} + {a}\\cdot{r**(a-1)}\\cdot x + \\text{{higher order terms}}$",
                f"$\\lim_{{x \\to 0}} \\frac{{{r**a} + {a*r**(a-1)}x + ... - {r**a}}}{{x}}$",
                f"$\\lim_{{x \\to 0}} \\frac{{{a*r**(a-1)}x + ...}}{{x}}$",
                f"$\\lim_{{x \\to 0}} ({a*r**(a-1)} + ...) = {ans}$"
            ),
        )

emit(generate())

File: src/calculus_limits_algebraic_hard.jl
# calculus - limits_algebraic (hard)
# Generated: 2026-03-08T20:36:58.129923

include(joinpath(@__DIR__, "..", "julia", "src", "ProblemUtils.jl"))
using .ProblemUtils

@script x h begin
    set_topic!("calculus/limits_algebraic")
    
    problem_type = rand(1:5)
    
    if problem_type == 1
        # Indeterminate form 0/0 requiring factoring
        # Target: 1800-2100
        r = nonzero(-12, 12)
        a = nonzero(-8, 8)
        b = randint(-15, 15)
        
        numerator = expand(a * (x - r) * (x + b))
        denominator = x - r
        
        limit_value = a * (r + b)
        
        problem(
            question="Evaluate \\lim_{x \\to $(r)} \\frac{$(tex(numerator))}{$(tex(denominator))}",
            answer=limit_value,
            difficulty=(1800, 2100),
            solution=steps(
                sol("Given", "\\lim_{x \\to $(r)} \\frac{$(tex(numerator))}{$(tex(denominator))}"),
                "Factor numerator: \$$(tex(a * (x - r) * (x + b)))\$",
                "Cancel common factor \$(x - $(r))\$",
                "Simplified: \$\\lim_{x \\to $(r)} $(tex(a * (x + b)))\$",
                "Substitute \$x = $(r)\$",
                sol("Answer", limit_value)
            ),
            time=150
        )
        
    elseif problem_type == 2
        # Rationalize conjugate - sqrt indeterminate form
        # Target: 1900-2300
        a = rand([1, 2, 4])
        c = randint(1, 20)
        
        # lim_{x->c} (sqrt(ax) - sqrt(ac)) / (x - c)
        # = lim_{x->c} (ax - ac) / [(x-c)(sqrt(ax) + sqrt(ac))]
        # = lim_{x->c} a(x-c) / [(x-c)(sqrt(ax) + sqrt(ac))]
        # = a / (2*sqrt(ac))
        
        limit_value_num = a
        limit_value_den = 2 * sqrt(a * c)
        
        if limit_value_den == round(limit_value_den)
            limit_value = limit_value_num // Int(limit_value_den)
        else
            limit_value = limit_value_num / limit_value_den
        end
        
        problem(
            question="Evaluate \\lim_{x \\to $(c)} \\frac{\\sqrt{$(a)x} - $(tex(sqrt(a*c)))}{x - $(c)}",
            answer=limit_value,
            difficulty=(1900, 2300),
            solution=steps(
                sol("Given", "\\lim_{x \\to $(c)} \\frac{\\sqrt{$(a)x} - $(tex(sqrt(a*c)))}{x - $(c)}"),
                "Multiply by conjugate: \$\\frac{\\sqrt{$(a)x} + $(tex(sqrt(a*c)))}{\\sqrt{$(a)x} + $(tex(sqrt(a*c)))}\$",
                "Numerator becomes: \$$(a)x - $(a*c) = $(a)(x - $(c))\$",
                "Cancel \$(x - $(c))\$ factor",
                "Simplified: \$\\lim_{x \\to $(c)} \\frac{$(a)}{\\sqrt{$(a)x} + $(tex(sqrt(a*c)))}\$",
                "Substitute \$x = $(c)\$",
                sol("Answer", limit_value)
            ),
            time=180
        )
        
    elseif problem_type == 3
        # Complex rational function - highest degree analysis
        # Target: 1850-2150
        a = nonzero(-6, 6)
        b = nonzero(-8, 8)
        c = nonzero(-5, 5)
        d = nonzero(-7, 7)
        
        numerator = a*x^2 + b*x + randint(-10, 10)
        denominator = c*x^2 + d*x + randint(-10, 10)
        
        limit_value = a // c
        
        problem(
            question="Evaluate \\lim_{x \\to \\infty} \\frac{$(tex(numerator))}{$(tex(denominator))}",
            answer=limit_value,
            difficulty=(1850, 2150),
            solution=steps(
                sol("Given", "\\lim_{x \\to \\infty} \\frac{$(tex(numerator))}{$(tex(denominator))}"),
                "Divide numerator and denominator by highest power \$x^2\$",
                "Leading coefficients: numerator has \$$(a)\$, denominator has \$$(c)\$",
                "Lower degree terms approach 0 as \$x \\to \\infty\$",
                sol("Answer", limit_value)
            ),
            time=120
        )
        
    elseif problem_type == 4
        # Difference of cubes or higher polynomial factoring
        # Target: 2000-2400
        r = nonzero(-8, 8)
        a = nonzero(-4, 4)
        
        # (x^3 - r^3)/(x - r) = x^2 + rx + r^2 at x=r
        numerator = x^3 - r^3
        denominator = x - r
        
        limit_value = r^2 + r*r + r^2
        limit_value = 3 * r^2
        
        problem(
            question="Evaluate \\lim_{x \\to $(r)} \\frac{x^3 - $(r^3)}{x - $(r)}",
            answer=limit_value,
            difficulty=(2000, 2400),
            solution=steps(
                sol("Given", "\\lim_{x \\to $(r)} \\frac{x^3 - $(r^3)}{x - $(r)}"),
                "Factor using difference of cubes: \$a^3 - b^3 = (a-b)(a^2 + ab + b^2)\$",
                "Numerator: \$(x - $(r))(x^2 + $(r)x + $(r^2))\$",
                "Cancel \$(x - $(r))\$ factor",
                "Evaluate: \$\\lim_{x \\to $(r)} (x^2 + $(r)x + $(r^2))\$",
                "Substitute \$x = $(r)\$: \$$(r^2) + $(r*r) + $(r^2) = $(limit_value)\$",
                sol("Answer", limit_value)
            ),
            time=200
        )
        
    else
        # Piecewise-motivated limit with algebraic simplification
        # Target: 2100-2500
        a = nonzero(-6, 6)
        b = nonzero(-10, 10)
        c = randint(1, 12)
        
        # |ax + b| / (ax + b) as x -> critical point from one side
        # Need ax + b = 0, so x = -b/a
        critical = -b // a
        
        if a > 0
            # From left: negative, limit = -1
            # From right: positive, limit = 1
            side = rand(["left", "right"])
            limit_value = side == "left" ? -1 : 1
            notation = side == "left" ? "^-" : "^+"
        else
            side = rand(["left", "right"])
            limit_value = side == "left" ? 1 : -1
            notation = side == "left" ? "^-" : "^+"
        end
        
        problem(
            question="Evaluate \\lim_{x \\to $(tex(critical))$(notation)} \\frac{|$(tex(a*x + b))|}{$(tex(a*x + b))}",
            answer=limit_value,
            difficulty=(2100, 2500),
            solution=steps(
                sol("Given", "\\lim_{x \\to $(tex(critical))$(notation)} \\frac{|$(tex(a*x + b))|}{$(tex(a*x + b))}"),
                "Note that \$$(tex(a*x + b)) = 0\$ when \$x = $(tex(critical))\$",
                "As \$x \\to $(tex(critical))$(notation)\$, the expression \$$(tex(a*x + b))\$ is $(side == "left" ? "negative" : "positive") (approaching from $(side))",
                "Therefore \$|$(tex(a*x + b))| = $(limit_value == 1 ? tex(a*x + b) : tex(-(a*x + b)))\$",
                "The ratio simplifies to \$$(limit_value)\$",
                sol("Answer", limit_value)
            ),
            time=180
        )
    end
end